src/examples/pmreorder/pmreorder_list.c

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright 2018, Intel Corporation
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in
+ *       the documentation and/or other materials provided with the
+ *       distribution.
+ *
+ *     * Neither the name of the copyright holder nor the names of its
+ *       contributors may be used to endorse or promote products derived
+ *       from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/*
+ * pmreorder_list.c -- explains how to use pmreorder tool with libpmem
+ *
+ * usage: pmreorder_list <g|b|c> <path>
+ * g - good case - add element to the list in a consistent way
+ * b - bad case - add element to the list in an inconsistent way
+ * c - check persistent consistency of the list
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <libpmem.h>
+
+#define MAX_NODES 10
+#define NODE_PTR(root, node_id) (&(root)->nodes[(node_id)])
+
+typedef size_t node_id;
+
+struct list_node {
+	int value;
+	node_id next;
+};
+
+struct list_root {
+	node_id head;
+	struct list_node nodes[MAX_NODES];
+};
+
+/*
+ * check_consistency -- check if list meets the set requirements
+ *
+ * for consistent cases function returns 0
+ * for inconsistent cases function returns 1
+ */
+static int
+check_consistency(struct list_root *root)
+{
+	/* if all nodes are set... */
+	if (NODE_PTR(root, 5)->value != 0 &&
+			NODE_PTR(root, 3)->value != 0 &&
+			NODE_PTR(root, 6)->value != 0) {
+		/* ...then we expect they have a proper order and values */
+		struct list_node *node1 = NODE_PTR(root, root->head);
+		struct list_node *node2 = NODE_PTR(root, node1->next);
+		struct list_node *node3 = NODE_PTR(root, node2->next);
+
+		if (node1->value != 66 ||
+			node2->value != 33 ||
+			node3->value != 55)
+			return 1;
+	}
```

Comment:
You are not verifying the "next" field and perform the check only when all values are filled - this is not very useful consistency check. I think this needs to be refactored to actually follow the stored list . See also my other comments.